### PR TITLE
🌐Bulk get GitHub commits

### PR DESCRIPTION
**Pull request by yufeih**
This PR fixes GitHub API abuse/rate limit and performance problem by:

- Using bulk API calls to get commits and cache users more aggressively (bulk API consumes the same rate as non-bulk API)
- Make GitHubUserCache serial, to avoid API abuse
- Make default cache expiration in 15 ~ 30 days

Test Result:

Repo | User profile cache | # of GitHub API calls | Build Time | Abused / rate limited?
------|--------------------------|------------------------|------------|-------
azure-docs-pr | no | 2091 |  00:26:50 | no
azure-docs-pr | yes (today's sandbox cache, may have expired records) | 210 | 00:03:36 | no
azure-docs-pr | yes (impact test cache, full) | 0 | 00:01:43 | no
edge-developer | no | 13 | 10s | no
edge-developer | yes (today's sandbox cache, may have expired records) | 1 | 3s | no

#4139

### CODE REVIEW

File: src/docfx/lib/github/GitHubAccessor.cs [hunk 1]
```diff
@@ -81,17 +78,15 @@ public async Task<(Error, GitHubUser)> GetUserByCommit(string repoOwner, string
             var apiDetail = $"GET /repos/{repoOwner}/{repoName}/commits/{commitSha}";
```
Comment (OsmondJiang) [inline src/docfx/lib/github/GitHubAccessor.cs]: is `apiDetail` need to be changed to new API?

File: src/docfx/lib/github/GitHubUserCache.cs [hunk 2]
```diff
@@ -154,55 +149,59 @@ async Task<(Error, GitHubUser)> GetByCommitCore()
...
                 // only mark the email as invalid when the user is not found
-                if (user != null)
+                if (users != null)
                 {
-                    UpdateUser(user);
+                    UpdateUsers(users);
                 }
-                return (error, user);
+
+                return (error, _usersByEmail.TryGetValue(authorEmail, out var user) && user.IsValid() ? user : null);
```
Comment (OsmondJiang) [inline src/docfx/lib/github/GitHubUserCache.cs]: will that cause any differences  between `get by commit` and `get by email`?

File: src/docfx/lib/github/GitHubUserCache.cs [hunk 3]
```diff
@@ -138,13 +133,13 @@ public Task<(Error error, GitHubUser user)> GetByCommit(string authorEmail, stri
             if (string.IsNullOrEmpty(authorEmail))
```
Comment (OsmondJiang) [inline src/docfx/lib/github/GitHubUserCache.cs]: does that mean if the author email is empty, we can't get any user from github even using commit?

Comment (yufeih) [reply src/docfx/lib/github/GitHubUserCache.cs]: yep

Comment (OsmondJiang) [reply src/docfx/lib/github/GitHubUserCache.cs]: we should be able to get user by login(got from commit) even if the author email is empty?

File: src/docfx/lib/github/GitHubUserCache.cs (hunk 2 above)
Comment (yufeih) [reply src/docfx/lib/github/GitHubUserCache.cs]: The `string.IsNullOrEmpty(repoOwner)` will make `get by email` return early, so there is no different. But it is bad to mix these into one method, I need to split them in another PR later.

**Pull request merged**